tapgarden/re-org_watcher.go

```diff
@@ -0,0 +1,565 @@
+	// backend.
+	ChainBridge ChainBridge
+
+	// ProofArchive is the storage backend for proofs to which we store
+	// updated proofs.
+	ProofArchive proof.NotifyArchiver
+
+	// SafeDepth is the number of confirmations we require before we
+	// consider a transaction to be safely buried in the chain.
+	SafeDepth int32
+
+	// ErrChan is the main error channel the watcher will report back
+	// critical errors to the main server.
+	ErrChan chan<- error
+}
+
+// ReOrgWatcher is responsible for watching initially confirmed transactions
+// until they reach a safe confirmation depth. If a re-org happens, it will
+// update the proof and store it in the proof archive.
+type ReOrgWatcher struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	cfg *ReOrgWatcherConfig
+
+	bestHeight atomic.Int32
+
+	incomingProofs chan *proofRegistration
+	incomingConfs  chan *chainntnfs.TxConfirmation
+
+	// pendingProofs is a list of all proofs that are currently being
+	// watched for re-orgs, keyed by their anchor transaction hash.
+	pendingProofs map[chainhash.Hash]*anchorTxNotification
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewReOrgWatcher creates a new re-org watcher based on the passed config.
+func NewReOrgWatcher(cfg *ReOrgWatcherConfig) *ReOrgWatcher {
+	return &ReOrgWatcher{
+		cfg:            cfg,
+		incomingProofs: make(chan *proofRegistration),
+		incomingConfs:  make(chan *chainntnfs.TxConfirmation),
+		pendingProofs:  make(map[chainhash.Hash]*anchorTxNotification),
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}
+}
+
+// Start attempts to start a new re-org watcher.
+func (w *ReOrgWatcher) Start(assets []*asset.Asset) error {
+	var startErr error
+	w.startOnce.Do(func() {
+		log.Info("Starting re-org watcher")
+
+		// Start the main event handler loop that will process new
+		// proofs and watch their anchor transactions until they reach
+		// a safe confirmation depth.
+		w.Wg.Add(1)
+		go w.watchTransactions()
+
+		// Now that we have started the watcher, we can load all assets
+		// that are not yet sufficiently deep buried and watch their
+		// anchor transactions for re-orgs.
+		ctx, cancel := w.WithCtxQuitNoTimeout()
+		defer cancel()
+		for idx := range assets {
+			locator := proof.Locator{
+				AssetID:   fn.Ptr(assets[idx].ID()),
+				ScriptKey: *assets[idx].ScriptKey.PubKey,
+			}
+			blob, err := w.cfg.ProofArchive.FetchProof(ctx, locator)
+			if err != nil {
+				startErr = fmt.Errorf("unable to fetch proof "+
+					"for asset %v: %w", assets[idx].ID(),
+					err)
+				return
+			}
+
+			f := &proof.File{}
+			err = f.Decode(bytes.NewReader(blob))
+			if err != nil {
+				startErr = fmt.Errorf("unable to decode "+
+					"proof: %w", err)
+				return
+			}
+
+			err = w.MaybeWatch(f, w.DefaultUpdateCallback())
```

Comment:
Re the earlier suggestion, if we only pass the items that actually are in "danger" of a re-org, then we wouldn't need this check. We'd also be able to avoid reading all the proof data above here as well (for all the open assets that is).

Reply:
I think even with the optimized query where we only fetch the assets not yet buried, we'll want to inspect the full proof file as there might be ancestor transactions that haven't fully confirmed yet.